uniffi_core/src/ffi/rustfuture/scheduler.rs

```diff
@@ -2,11 +2,72 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
-use std::mem;
+use std::{cell::RefCell, future::poll_fn, mem, task::Poll, thread_local};
 
 use super::{RustFutureContinuationCallback, RustFuturePoll};
 use crate::Handle;
 
+/// Context of the current `RustFuture::poll` call
+struct RustFutureContext {
+    /// Blocking task queue that the future is being polled on
+    current_blocking_task_queue: Option<Handle>,
+    /// Blocking task queue that we've been asked to schedule the next poll on
+    scheduled_blocking_task_queue: Option<Handle>,
+}
+
+thread_local! {
+    static CONTEXT: RefCell<RustFutureContext> = RefCell::new(RustFutureContext {
+        current_blocking_task_queue: None,
+        scheduled_blocking_task_queue: None,
+    });
+}
+
+fn with_context<F: FnOnce(&mut RustFutureContext) -> R, R>(operation: F) -> R {
+    CONTEXT.with(|context| operation(&mut *context.borrow_mut()))
+}
+
+/// Schedule work in a blocking task queue
+///
+/// The returned future will attempt to arrange for [RustFuture::poll] to be called in the
+/// blocking task queue.  Once [RustFuture::poll] is running in the blocking task queue, then the future
+/// will be ready.
+///
+/// There's one tricky issue here: how can we ensure that when the top-level task is run in the
+/// blocking task queue, this future will be polled?  What happens this future is a child of `join!`,
+/// `FuturesUnordered` or some other Future that handles its own polling?
+///
+/// We start with an assumption: if we notify the waker then this future will be polled when the
+/// top-level task is polled next.  If a future does not honor this then we consider it a broken
+/// future.  This seems fair, since that future would almost certainly break a lot of other future
+/// code.
```

Comment:
I'm pretty sure this is a good assumption, but I'd love a double check on this one.

When I mention "other future code", I was thinking of something like a lock.  If your async lock type locks a mutex and notifies the waker, then any intermediate futures must poll your type or else you're in a bad state.  Unless I'm missing something, async lock authors just have assume they will get polled and we can too.